docs/internals/task_lifecycle.md

```diff
@@ -0,0 +1,186 @@
+task sub-types in Flink's streaming engine. 
+
+* This will be replaced by the TOC
+{:toc}
+
+## Operator Lifecycle in a nutshell
+
+Because the task is the entity that executes your operators, its lifecycle is tightly integrated with that of an operator. 
+So, we will briefly mention the basic methods representing the lifecycle of an operator before diving into those of the 
+`StreamTask` itself. The list is presented below in the order that each of the methods is called. Given that an operator
+can have a user-defined function (*UDF*), below each of the operator methods we also present (indented) the methods in 
+the lifecycle of UDF that it calls. These methods are available if your operator extends the `AbstractUdfStreamOperator`, 
+which is the basic class for all operators that execute UDFs.
+
+        // initialization
+        OPERATOR::setup
+            UDF::setRuntimeContext
+        OPERATOR::initializeState
+        OPERATOR::open
+            UDF::open
+        
+        // processing
+        OPERATOR::processElement
+            UDF::run
+        OPERATOR::processWatermark
+        
+        // checkpointing (called asynchronously)
+        OPERATOR::snapshotState
+                
+        // termination
+        OPERATOR::close
+            UDF::close
+        OPERATOR::dispose
+    
+In a nutshell, the `setup()` is called to initialize some operator-specific machinery, such as its `RuntimeContext` and 
+its metric collection data-structures. After this, the `initializeState()` gives an operator its initial state, and the 
+ `open()` method executes any operator-specific initialization, such as opening the user-defined function in the case of 
+the `AbstractUdfStreamOperator`. 
+
+<span class="label label-danger">Attention</span> The `initializeState()` contains both the logic for initializing the 
+state of the operator during its initial execution (*e.g.* register any keyed state), and also the logic to retrieve its
+state from a checkpoint after a failure. More about this on the rest of this page.
+
+Now that everything is set, the operator is ready to process incoming data. This is done by invoking the `processElement()` 
+and `processWatermark()` methods which contain the logic for processing elements and watermarks, respectively. The 
+`processElement()` is also the place where you function's logic is invoked, *e.g.* the `map()` method of your `MapFunction`.
+
+Finally, in the case of a normal, fault-free termination of the operator (*e.g.* if the stream is finite and its end is 
+reached), the `close()` method is called to perform any final bookkeeping action required by the operator's logic, and 
+the `dispose()` is called after that to free any resources held by the operator. 
+
+In the case of a termination due to a failure or due to manual cancellation, the execution jumps directly to the `dispose()` 
```

Comment:
This is also why failure go immediately to `dispose`: the operator is already stopped (because of the failure), but the resources are not yet all released.